UnitySDK/Assets/ML-Agents/Scripts/Grpc/RpcCommunicator.cs

```diff
@@ -217,42 +227,53 @@ void SendCommandEvent(CommandProto command)
 
         public void DecideBatch()
         {
-            if (m_CurrentAgents.Values.All(l => l.Count == 0))
+            if (!m_NeedCommunicateThisStep)
             {
                 return;
             }
-            foreach (var brainKey in m_CurrentAgents.Keys)
+            m_NeedCommunicateThisStep = false;
+
+            SendBatchedMessageHelper();
+        }
+
+        /// <summary>
+        /// Sends the observations of one Agent. 
+        /// </summary>
+        /// <param name="brainKey">Batch Key.</param>
+        /// <param name="agent">Agent info.</param>
+        public void PutObservations(string brainKey, AgentInfo info, List<ISensor> sensors, Action<AgentAction> action)
+        {
+            using (TimerStack.Instance.Scoped("AgentInfo.ToProto"))
             {
-                using (TimerStack.Instance.Scoped("AgentInfo.ToProto"))
+
+                if (!m_ActionCallbacks.ContainsKey(brainKey))
                 {
-                    if (m_CurrentAgents[brainKey].Count > 0)
+                    int numFloatObservations = 0;
+                    for (var i = 0; i < sensors.Count; i++)
                     {
-                        foreach (var agent in m_CurrentAgents[brainKey])
+                        if (sensors[i].GetCompressionType() == SensorCompressionType.None)
                         {
-                            // Update the sensor data on the AgentInfo
-                            agent.GenerateSensorData();
-                            var agentInfoProto = agent.Info.ToAgentInfoProto();
-                            m_CurrentUnityRlOutput.AgentInfos[brainKey].Value.Add(agentInfoProto);
+                            numFloatObservations += sensors[i].ObservationSize();
                         }
+                    }
 
+                    if (m_VectorObservationBuffer.Length < numFloatObservations)
```

Comment:
This is problematic if we try to use the Observation anywhere outside of here, because each observation will be referring to the same buffer. The protos will be fine, but all the float values will be overwritten by subsequent Agents' sensors.

Reply:
Isn't that what we want? Reuse the buffer but have the right proto being sent ?

Reply:
But it means the AgentInfo.observations are dirty as soon another Agent computes their observations

Reply:
We shouldn't be leaving things in a bad state; I think removing AgentInfo.observations is a decent fix for this. If AgentInfo.observations are actually used in lots of places, this change will break them

Reply:
Ok, I made the change. Let me know what you think